### PR TITLE
Cli

### PR DESCRIPTION
**Pull request by Spill-Tea**
Convert of CLI to use a Typer Application, and begin to carve out future CLI options / commands / arguments.

### CODE REVIEW

**Comment by Spill-Tea** [comment]
For reference, this is the default help menu display:

```bash
fhy --help
```
Screenshot 2024-06-03 at 7 45 12 PM

And for the subcommand
```bash
fhy serialize --help
```
Screenshot 2024-06-03 at 7 45 36 PM


Using an environment variable, we can modify the help menu display to use rich text formatting as shown:

Screenshot 2024-06-03 at 7 52 36 PM

Screenshot 2024-06-03 at 7 52 55 PM

File: src/fhy/cli.py [hunk 1]
```diff
@@ -58,135 +98,201 @@ class Status(IntEnum):
     USAGE_ERROR = 3
 
 
-def make_logger(verbose: bool = False) -> logging.Logger:
-    """Constructs a Simple Logger."""
-    level: int = logging.DEBUG if verbose else logging.INFO
-    log: logging.Logger = get_logger("FhY", level=level)
+@dataclass
+class Namespace:
```
Comment (christopherpriebe) [inline src/fhy/cli.py:102]: What is a namespace here? Our language has namespaces like Python does. This is confusing to me.

File: src/fhy/cli.py [hunk 2]
```diff
@@ -58,135 +98,201 @@ class Status(IntEnum):
...
-    if args.log_file:
-        add_file_handler(log, args.log_file)
-        log.debug("Added File Handler to log instance: %s", args.log_file)
+@app.callback(
+    invoke_without_command=True,
+    context_settings={"allow_extra_args": True, "ignore_unknown_options": True},
+)
+def main(
+    ctx: typer.Context,
+    module: Annotated[
```
Comment (christopherpriebe) [inline src/fhy/cli.py:136]: I don't like the --module vs. --library command like arguments. I think that the keyword arguments specify compilation options. A user should be able to compile a FhY file without having to use any keyword arguments. The CLI should take in ONLY the root file containing the main procedure and compile from there.

File: src/fhy/lang/converter/from_parse_tree.py [hunk 3]
```diff
@@ -48,23 +48,25 @@
...
     if all((start, stop)):
-        return Span(start.line, stop.line, start.column, stop.column)
+        return Span(start.line, stop.line, start.column, stop.column, source)
 
     elif not parent and (parent_ctx := getattr(ctx, "parentCtx", None)) is not None:
-        return _get_source_info(parent_ctx, True)
+        return _get_source_info(parent_ctx, source, True)
 
-    return Span(0, 0, 0, 0)
+    return Span(0, 0, 0, 0, source)
```
Comment (christopherpriebe) [inline src/fhy/lang/converter/from_parse_tree.py:69]: Span should be an optional field in the AST node and then this should return None here.

**Review by christopherpriebe — CHANGES_REQUESTED** [review]

File: src/fhy/cli.py [hunk 4]
```diff
@@ -31,162 +31,281 @@
...
         sys.stdout.write(f"FhY v{__version__}\n")
-        return Status.OK
+        sys.exit(Status.OK)
+
+
+def _confirm_arg(value: bool, check: str) -> bool:
+    return value or check in sys.argv
+
+
+def compiles(
```
Comment (christopherpriebe) [inline src/fhy/cli.py:146]: Is there a better name that uses an active phrase rather than a passive one?

File: requirements.txt [hunk 5]
```diff
@@ -1 +1,2 @@
 networkx>=3.3
+typer
```
Comment (christopherpriebe) [inline requirements.txt:2]: change this to typer-slim. I think we should make the full typer install an optional dependency rather than the default

**Review by christopherpriebe — CHANGES_REQUESTED** [review]
Some small things but overall it looks good.

Oh, also, if you can make the "usage" part of the help menu make more sense that would be awesome because right now it just says "Usage: fhy [OPTIONS] COMMAND [ARGS]..." instead of something more useful like "Usage: fhy [OPTIONS] MAIN_FILE.fhy" or something idk

**Comment by christopherpriebe** [comment]
lgtm!

**Review by christopherpriebe — APPROVED** [review]
lgtm!

**Pull request merged**